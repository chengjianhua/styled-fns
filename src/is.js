import {css} from 'styled-components'
import invariant from 'invariant'

export default function is(checker) {
  const type = typeof checker

  invariant(
    type === 'string' || type === 'function',
    '[styled-fns]: `is()` should accept string or function as argument',
  )

  return (...args) => props => {
    if (typeof checker === 'function') {
      return checker(props) ? css(...args) : null
    }

    return props[checker] ? css(...args) : null
  }
}
