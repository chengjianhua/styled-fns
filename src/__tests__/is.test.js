import React from 'react'
import renderer from 'react-test-renderer'
import styled from 'styled-components'
// eslint-disable-next-line import/no-unassigned-import
import 'jest-styled-components'

import is from '../is'

describe('is', () => {
  it('should throw error if argument is not type of string or function', () => {
    expect(() => {
      // eslint-disable-next-line no-unused-vars
      const Comp = styled.div`
        ${is(true)``};
      `
    }).toThrowErrorMatchingSnapshot()
  })

  it('should accept a string as argument `name` to check whether `props[name]` is true or false', () => {
    // const checker = is('loading')`colors`;
    const Example = styled.div`
      ${is('loading')`
        display: none;
      `};
    `

    const tree1 = renderer.create(<Example loading />).toJSON()

    expect(tree1).toHaveStyleRule('display', 'none')
    expect(tree1).toMatchSnapshot()

    const tree2 = renderer.create(<Example />).toJSON()

    expect(tree2).not.toHaveStyleRule('display', 'none')
    expect(tree2).toMatchSnapshot()
  })

  it('should accept a function as argument `checker` to check whether `checker(props) is true or false`', () => {
    const Example = styled.div`
      ${is(({loading, primary}) => loading && primary)`
        display: none;
      `};
    `

    const tree1 = renderer.create(<Example loading primary />).toJSON()

    expect(tree1).toHaveStyleRule('display', 'none')
    expect(tree1).toMatchSnapshot()

    const tree2 = renderer.create(<Example loading />).toJSON()

    expect(tree2).not.toHaveStyleRule('display', 'none')
    expect(tree2).toMatchSnapshot()

    const tree3 = renderer.create(<Example primary />).toJSON()

    expect(tree3).not.toHaveStyleRule('display', 'none')
    expect(tree3).toMatchSnapshot()
  })
})
