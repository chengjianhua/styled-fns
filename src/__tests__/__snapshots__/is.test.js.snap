// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is should accept a function as argument \`checker\` to check whether \`checker(props) is true or false\` 1`] = `
.c0 {
  display: none;
}

<div
  className="c0"
/>
`;

exports[`is should accept a function as argument \`checker\` to check whether \`checker(props) is true or false\` 2`] = `
<div
  className=""
/>
`;

exports[`is should accept a function as argument \`checker\` to check whether \`checker(props) is true or false\` 3`] = `
<div
  className=""
/>
`;

exports[`is should accept a string as argument \`name\` to check whether \`props[name]\` is true or false 1`] = `
.c0 {
  display: none;
}

<div
  className="c0"
/>
`;

exports[`is should accept a string as argument \`name\` to check whether \`props[name]\` is true or false 2`] = `
<div
  className=""
/>
`;

exports[`is should throw error if argument is not type of string or function 1`] = `"[styled-fns]: \`is()\` should accept string or function as argument"`;
